pkg/deploy/strategy/support/lifecycle.go

```diff
@@ -203,3 +203,121 @@ func NewPodWatch(client kclient.Interface, namespace, name, resourceVersion stri
+	// For now, only validate the first replica.
+	if deployment.Spec.Replicas != 1 {
+		return nil
+	}
+
+	// Try and find the pod for the deployment.
+	pods, err := c.podsForDeployment(deployment)
+	if err != nil {
+		return fmt.Errorf("couldn't get pods for deployment %s: %v", deployutil.LabelForDeployment(deployment), err)
+	}
+	if len(pods.Items) == 0 {
+		return fmt.Errorf("no pods found for deployment %s", deployutil.LabelForDeployment(deployment))
+	}
+
+	// If we found multiple, use the first one and log a warning.
+	// TODO: should finding multiple be an error?
+	pod := pods.Items[0]
+	if len(pods.Items) > 0 {
+		glog.Infof("Warning: more than one pod for deployment %s; basing canary check on the first pod '%s'", deployutil.LabelForDeployment(deployment), pod.Name)
+	}
+
+	// Wait for container readiness to be observed once per container in the pod.
+	containersReady := map[string]bool{}
+	for _, container := range pod.Spec.Containers {
+		containersReady[container.Name] = false
+	}
+
+	// Make a pod store to poll and ensure it gets cleaned up.
+	podStore, stopStore := c.getPodStore(pod.Namespace, pod.Name)
+	defer close(stopStore)
+
+	// Set up the overall timeout and the poll period.
+	timeout := time.After(c.timeout)
+	poll := time.NewTicker(1 * time.Second)
+	defer poll.Stop()
+
+	// Start waiting for the pod.
+	glog.V(0).Infof("Waiting for pod %s/%s container readiness", pod.Namespace, pod.Name)
+	for {
+		select {
+		case <-timeout:
+			return fmt.Errorf("timeout exceeded waiting for pod %s/%s container readiness; observed container statuses: #%v", pod.Namespace, pod.Name, containersReady)
+		case <-poll.C:
+			// Get the latest state of the pod.
+			obj, exists, err := podStore.Get(pod)
+			if !exists {
+				glog.V(0).Infof("Couldn't find pod %s/%s to inspect container readiness", pod.Namespace, pod.Name)
+				break
+			}
+			if err != nil {
+				glog.V(0).Infof("Error getting pod %s/%s to inspect container readiness: %v", pod.Namespace, pod.Name, err)
+				break
+			}
+			updatedPod := obj.(*kapi.Pod)
+			// Update the observed ready status of any containers.
+			for _, status := range updatedPod.Status.ContainerStatuses {
+				if _, known := containersReady[status.Name]; !known {
+					glog.V(0).Infof("Ignoring readiness of container %s in pod %s/%s because it's not present in the pod spec", status.Name, pod.Namespace, pod.Name)
+					continue
+				}
+				// The status of the container could be transient; we only care if it
+				// was ever ready. If it was ready and then became not ready, we
+				// consider it ready.
+				if status.Ready {
+					containersReady[status.Name] = true
+				}
+			}
+			// Determine if we're done.
+			allReady := true
+			for _, ready := range containersReady {
+				if !ready {
+					allReady = false
+					break
+				}
+			}
+			// If all containers have been ready once, return success.
+			if allReady {
+				glog.V(0).Infof("All containers ready for %s/%s", pod.Namespace, pod.Name)
```

Comment:
The only time this code is used is via the deployer code running in a pod (the logs are only accessible right now via `docker log` and the log level isn't configurable). The other deployer code doesn't even use log levels for this reason. Should I continue to omit levels here for the same reason, or what? I really want to see this stuff from the container logs.